framework/base/Controller.php

```diff
@@ -63,6 +63,28 @@ class Controller extends Component implements ViewContextInterface
      * by [[run()]] when it is called by [[Application]] to run an action.
      */
     public $action;
+    /**
+     * @var array middleware stack to be applied for this controller. Particular array element
+     * represents a particular middleware instance or its DI compatible configuration.
+     * For example:
+     *
+     * ```php
+     * [
+     *     new SomeMiddleware(),
+     *     [
+     *         'class' => AnotherMiddleware::class,
+     *         'someProperty' => 'foo',
+     *     ],
+     *     function () {
+     *         return (new SomeMiddleware())->withParams(['param1' => '1']);
+     *     }
+     * ],
+     * ```
+     *
+     * @see MiddlewareDispatcherInterface::dispatch()
+     * @since 2.1.0
+     */
+    public $middleware = [];
```

Comment:
Middleware is meant to be set on the router/url manager config level so it could be applied to a number of routes some of which could be non-controllers. It's not meant to be part of the controller itself although previously filters were controller-only thing.